pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,494 @@
+
+	klog.Infof("Adding %s to electionQueue", leaderLeaseID)
+	c.electionQueue.Add(leaderLeaseID)
+}
+
+func (c *Controller) processLease(obj any) error {
+	l, ok := obj.(*v1.Lease)
+	if !ok {
+		return nil
+	}
+	key, err := controller.KeyFunc(l)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", l, err))
+	}
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	lease, err := c.leaseInformer.Lister().Leases(namespace).Get(name)
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if lease != nil {
+		// If the lease is not held by a CLE aware holder and the lease is not expired
+		// then PreferredHolder is not recognized and election is unnecessary since
+		// the holder will not voluntarily give up the lease.
+		// This may occur on version skew cases when CLE is first enabled.
+		if lease.Annotations[ElectedByAnnotationName] != controllerName && !isLeaseExpired(lease) {
+			return nil
+		}
+	}
+
+	c.electionQueue.Add(leaderLeaseID{namespace: namespace, name: name})
+	return nil
+}
+
+func (c *Controller) electionNeeded(candidates []*v1alpha1.LeaseCandidate, leaderLeaseID leaderLeaseID) (bool, error) {
+	lease, err := c.leaseInformer.Lister().Leases(leaderLeaseID.namespace).Get(leaderLeaseID.name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("Error reading lease")
+	}
+	if len(candidates) == 0 {
+		return true, nil
+	}
+
+	prelimStrategy := pickBestStrategy(candidates)
+	if prelimStrategy != v1.OldestEmulationVersion {
+		return false, fmt.Errorf("strategy %s is not recognized by CLE", prelimStrategy)
+	}
+
+	prelimElectee := pickBestLeader(candidates)
+	if lease != nil && lease.Spec.HolderIdentity != nil && prelimElectee.Name == *lease.Spec.HolderIdentity {
+		klog.Infof("Leader %s is already most optimal for lease %s %s", prelimElectee.Name, leaderLeaseID.namespace, leaderLeaseID.name)
+		return false, nil
+	}
+	return true, nil
+}
+
+// stepElection steps through a step in an election.
+// A step can be defined as one of
+// - do nothing (because leader is already optimal or still waiting for an event)
+// - request ack from candidates
+// - finds the most optimal candidate and elect
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + 5 seconds > time.Now: We just asked all candidates to ack and are still waiting for response
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, proceed with election
+// The missing case is that the election should auto proceed after 5 seconds even if we don't get all acks.
+// I think we should be able to find the right state with the information available...
+func (c *Controller) stepElection(ctx context.Context, leaderLeaseID leaderLeaseID) error {
+	now := time.Now()
+	if leaderLeaseID.name == controllerName {
```

Comment:
is the name field the one that has the controller name or the annotation?

Reply:
filtering is already done in the processLease section so this is not necessary anymore.